datafusion/src/optimizer/constant_folding.rs

```diff
@@ -127,164 +124,365 @@ impl<'a> ConstantRewriter<'a> {
+                            self.const_fold_list_eager(args);
+                            false
                         }
-                        _ => Expr::ScalarFunction {
-                            fun: BuiltinScalarFunction::ToTimestamp,
-                            args,
-                        },
                     }
+                }
+            }
+            Expr::ScalarUDF { fun, args } => {
+                if args.is_empty() {
+                    false
                 } else {
-                    Expr::ScalarFunction {
-                        fun: BuiltinScalarFunction::ToTimestamp,
-                        args,
+                    let volatility = fun.volatility();
+                    match volatility {
+                        Volatility::Immutable => self.const_fold_list(args),
+                        _ => {
+                            self.const_fold_list_eager(args);
+                            false
+                        }
                     }
                 }
             }
-            Expr::Cast {
-                expr: inner,
-                data_type,
-            } => match inner.as_ref() {
-                Expr::Literal(val) => {
-                    let scalar_array = val.to_array();
-                    let cast_array = kernels::cast::cast_with_options(
-                        &scalar_array,
-                        &data_type,
-                        &DEFAULT_CAST_OPTIONS,
-                    )?;
-                    let cast_scalar = ScalarValue::try_from_array(&cast_array, 0)?;
-                    Expr::Literal(cast_scalar)
+            Expr::AggregateFunction { args, .. } => {
+                self.const_fold_list_eager(args);
+                false
+            }
+            Expr::WindowFunction {
+                args,
+                partition_by,
+                order_by,
+                ..
+            } => {
+                self.const_fold_list_eager(args);
+                self.const_fold_list_eager(partition_by);
+                self.const_fold_list_eager(order_by);
+                false
+            }
+            Expr::AggregateUDF { args, .. } => {
+                self.const_fold_list_eager(args);
+                false
+            }
+            Expr::InList { expr, list, .. } => {
+                let expr_const = self.rewrite_const_expr(expr);
+                let list_literals = self.const_fold_list(list);
+                match (expr_const, list_literals) {
+                    (true, true) => true,
+
+                    (false, false) => false,
+                    (true, false) => {
+                        if let Ok(s) = evaluate_const_expr_unchecked(expr) {
+                            let expr: &mut Expr = expr;
+                            *expr = Expr::Literal(s);
+                        }
+                        false
+                    }
+                    (false, true) => {
+                        self.const_fold_list_eager(list);
```

Comment:
In this case we run `rewrite_const_expr()` twice on each expression, is that expected?

Reply:
Well rereading the code it's expected but not intended haha.